### PR TITLE
Processors refactoring

### PR DESCRIPTION
**Pull request by Sergio0694**
### Prerequisites

- [X] I have written a descriptive pull-request title
- [X] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [X] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [X] I have provided test coverage for my change (where applicable)

### Description
This does some ground work for #967.
I'm creating this as a draft PR as I'm unsure about a couple of things, and there are a few lines that cause the current branch to fail to build. I've done most of the refactoring following step by step the instructions for the "Proposal to refactor current ImageProcessor logic" section.
Need some feedback to continue with this, and of course everyone else is free to just jump in and finish the remaining details as well 😄

#### What I did so far

- [X] Added `Image<TPixel> Source` and `Rectangle SourceRectangle` as `protected` properties to the `ImageProcessor<TPixel>` class
- [X] Also added a `Configuration` property (retrieved from the image)
- [X] Modified the `ImageProcessor<TPixel>` constructor to take an image and rectangle
- [X] Refactored constructors of all inherited processors to take and bubble up the image and rectangle
- [X] Modified `CreatePixelSpecificProcessor` to take target image and rectangle
- [X] Removed parameters from the `Apply` method, as they're now members
- [X] Did the same for the `CloningImageProcessor<TPixel>` and inherited processors
- [X] Removed duplicate parameters from all `ImageProcessor<TPixel>` methods (`Apply`, `BeforeApply`, `AfterFrameApply` and all the others)
- [X] Refactored the `DefaultImageProcessingContext` class to use the new architecture

### CODE REVIEW

**Comment by JimBobSquarePants** [comment]
@Sergio0694 I'll do some work on this later to get it building. Then we'll see where we are.

It'll be wise to add a checkbox list to the description here to tick off what you have done so far.

**Comment by Sergio0694** [comment]
Hey @JimBobSquarePants - I've just updated the PR description with all the info you requested, hope it helps! 😊

**Comment by JimBobSquarePants** [comment]
@Sergio0694 I got the source building, test still not yet though. If you could get them building I could then crack on with implementing `IDIsposable`.

File: src/ImageSharp/Processing/Processors/CloningImageProcessor.cs [hunk 1]
```diff
@@ -15,32 +15,59 @@ namespace SixLabors.ImageSharp.Processing.Processors
...
 
-                if (clone.Frames.Count != source.Frames.Count)
+                if (clone.Frames.Count != this.Source.Frames.Count)
                 {
                     throw new ImageProcessingException($"An error occurred when processing the image using {this.GetType().Name}. The processor changed the number of frames.");
                 }
 
-                Configuration configuration = source.GetConfiguration();
-                this.BeforeImageApply(source, clone, sourceRectangle);
+                Configuration configuration = this.Source.GetConfiguration();
```
[the remark this answers is not in the document]
Comment (Sergio0694) [reply src/ImageSharp/Processing/Processors/CloningImageProcessor.cs]: Fixed in https://github.com/SixLabors/ImageSharp/pull/983/commits/88e97cb4e59ad68daca7420bf83d6f820b6c32c4.

**Comment by Sergio0694** [comment]
@antonfirsov alright, I've fixed that small leftover you mentioned and marked the PR as ready for review, so you're free to merge this now if you want 👍

**Comment by JimBobSquarePants** [comment]
Ok, let’s merge this as-is then and I’ll open a new PR for using IDisposable since I’ve already made a start.

**Pull request merged**